core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -769,6 +800,75 @@ private SubPlan optimizePlan(SubPlan plan)
         {
             // Re-optimize plan here based on available runtime statistics.
             // Fragments changed due to re-optimization as well as their downstream stages are expected to be assigned new fragment ids.
+            plan = updateStagesPartitioning(plan);
+            return plan;
+        }
+
+        private SubPlan updateStagesPartitioning(SubPlan plan)
+        {
+            if (!runtimeAdaptivePartitioningEnabled || runtimeAdaptivePartitioningApplied) {
+                return plan;
+            }
+
+            for (SubPlan subPlan : planInTopologicalOrder) {
+                PlanFragment fragment = subPlan.getFragment();
+                if (!consumesHashPartitionedInput(fragment)) {
+                    // no input hash partitioning present
+                    continue;
+                }
+
+                StageId stageId = getStageId(fragment.getId());
+                if (stageExecutions.containsKey(stageId)) {
+                    // already started
```

Comment:
Is the assumption that stage which requires repartitining will never be started speculatively?

Reply:
OK so this review comment is not consistent with the code:) If a stage has already started, then we will not be able to insert repartition stages before it.

Reply:
Yeah - I know - that is why I ask if that is not a problem. That is can we end up not being able to fix the query (bump partition count) because some stage already started executing speculatively?

Reply:
That's not possible since we always check for input data size of a stage before it's started, and this applies to speculatively started stages as well.

You are probably asking stages whose parents were speculatively started. In that case we use the extrapolated estimated output data size to determine whether we want to change partition count at runtime.